// kmodel/index.js
var mongoose = require("mongoose"),
    Schema = mongoose.Schema;

exports.connect = function(url){
    if(this.connected) return this;
    mongoose.connect(url||'mongodb://localhost:27017/test');
    this.connected = true;
    return this;
};

// Model base-type
var Type = {
    
    string: String,

    number: Number,

    date:  Date,

    boolean: Boolean,

    buffer: Buffer,

    object: {},

    array: Array

};

var mixin = function(base, sup, option){

    if(option) mixin(base, option);

    for(var name in sup)
        base[name] = sup[name];

    return base;
};

var extend = function(sub, sup){
    var f = new Function();
    f.prototype = sup.prototype;
    sub.prototype = new f();
    sub.prototype.constructor = f.prototype;
};

var Class = function(model){
    this.model = model;
};

// 创建 Model
var createSchema = function(format, modelname){

    var schemas = {}, type;

    for(var key in format){
        type = format[key];
        if(type == undefined) continue;
        schemas[key] = Type[type];
    }

    schemas.ctime = Date;
    schemas.utime = Date;

    var schema = new Schema(schemas),
        model  = mongoose.model(modelname, schema);

    return model;
};

Class.prototype = {

    insertOne: function(instance, callback){
        instance.ctime = Date.now();
        instance.utime = Date.now();
        var m = new this.model(instance);
        m.save(callback);
    },

    updateOne: function(instance, option, callback){

        var model = this.model;

        if(!option) option = {_id: instance.id};
        instance.utime = Date.now();
        model.findOneAndUpdate(option, instance, callback);
    },

    deleteOne: function(option, callback){
        this.model.findOneAndRemove(option, callback);
    },

    queryOne: function(option, callback){
        this.model.findOne(option).exec(callback);
    },

    query: function(option, callback){
        this.model.find(option).exec(callback);
    }
};


exports.create = function(format, name){
    
    var m = createSchema(format, name),
        instance = new Class();
    
    instance.model = m;
    
    return instance;
};

exports.extend = function(instance, name, handler){
    instance[name] = handler;
};

exports.callback = function(callback){
    return function(err, instance){
        callback(instance || err);
    };
};

exports.all = function(name,  handler){
    Class.prototype[name] = handler;
};

exports.load = function(name){

    var f = this.loaded[name];

    if(f) return f;

    var handler = require("./features/"+name);

    if(!handler) console.error("feature not found");

    this.loaded[name] = handler[name](this);

    return this.loaded[name];
};

exports.loaded = {};
exports.Class = Class;
